include/swift/AST/DiagnosticEngine.h

```diff
@@ -802,29 +802,11 @@ namespace swift {
 
     bool isPrettyPrintingDecl() const { return IsPrettyPrintingDecl; }
 
-    void setLocalization(std::string locale, std::string path) {
+    void setLocalization(StringRef locale, StringRef path) {
       assert(!locale.empty());
       assert(!path.empty());
-      llvm::SmallString<128> filePath(path);
-      llvm::sys::path::append(filePath, locale);
-      llvm::sys::path::replace_extension(filePath, ".db");
-
-      // If the serialized diagnostics file not available,
-      // fallback to the `YAML` file.
-      if (llvm::sys::fs::exists(filePath)) {
-        if (auto file = llvm::MemoryBuffer::getFile(filePath)) {
-          localization = std::make_unique<diag::SerializedLocalizationProducer>(
-              std::move(file.get()), getPrintDiagnosticNames());
-        }
-      } else {
-        llvm::sys::path::replace_extension(filePath, ".yaml");
-        // In case of missing localization files, we should fallback to messages
-        // from `.def` files.
-        if (llvm::sys::fs::exists(filePath)) {
-          localization = std::make_unique<diag::YAMLLocalizationProducer>(
-              filePath.str(), getPrintDiagnosticNames());
-        }
-      }
+      localization = diag::LocalizationProducer::producerFor(
+          locale, path, getPrintDiagnosticNames());
```

Comment:
I moved the logic which was inside of `setLocalization` to the localization producer itself. I think it makes sense to add it there. WDYT?